language/diem-framework/modules/0L/Jail.move

```diff
@@ -0,0 +1,141 @@
+// It is recursive. If Alice is vouching for Carol, and Carol gets a Jail term. Then both Alice, and Bob will have lifetime_vouchees_jailed incremented. Even if Carol and Bob never met.
+// Not included in this code, but included in Vouch, is the logic for the deposit the Vouchers put on the Vouchee accounts. So besides the permanent mark on the account, Bob will lose the deposit.
+
+
+address 0x1 {
+  module Jail {
+    use 0x1::CoreAddresses;
+    use 0x1::TowerState;
+    use 0x1::Signer;
+    use 0x1::Vector;
+    use 0x1::Vouch;
+
+    struct Jail has key {
+        is_jailed: bool,
+        // number of times the validator was dropped from set. Does not reset.
+        lifetime_jailed: u64,
+        // number of times a downstream validator this user has vouched for has been jailed.
+        // this is recursive. So if a validator I vouched for, vouched for a third validator that failed, this number gets incremented.
+        lifetime_vouchees_jailed: u64,
+        // validator that was jailed and qualified to enter the set, but fails to complete epoch.
+        // this resets as soon as they rejoin successfully.
+        // this counter is used for ordering prospective validators entering a set.
+        consecutive_failure_to_rejoin: u64,
+
+    }
+
+    public fun init(val_sig: &signer) {
+      let addr = Signer::address_of(val_sig);
+      if (!exists<Jail>(addr)) {
+        move_to<Jail>(val_sig, Jail {
+          is_jailed: false,
+          lifetime_jailed: 0,
+          lifetime_vouchees_jailed: 0,
+          consecutive_failure_to_rejoin: 0,
+
+        });
+      }
+    }
+
+
+    public fun is_jailed(validator: address): bool acquires Jail {
+      if (!exists<Jail>(validator)) {
+        return false
+      };
+      borrow_global<Jail>(validator).is_jailed
+    }
+
+    public fun jail(vm: &signer, validator: address) acquires Jail{
+      CoreAddresses::assert_vm(vm);
+
+      assert(exists<Jail>(validator), 220101014011);
+
+      let j = borrow_global_mut<Jail>(validator);
+      j.is_jailed = true;
+      j.lifetime_jailed = j.lifetime_jailed + 1;
+      j.consecutive_failure_to_rejoin = j.consecutive_failure_to_rejoin + 1;
+    }
+
+    public fun remove_consecutive_fail(vm: &signer, validator: address) acquires Jail{
+      CoreAddresses::assert_vm(vm);
+      let j = borrow_global_mut<Jail>(validator);
+      j.consecutive_failure_to_rejoin = 0;
+
+    }
+
+    public fun vouch_unjail(sender: &signer, addr: address) acquires Jail {
+      // only a validator can un-jail themselves.
+      let voucher = Signer::address_of(sender);
+
+      let buddies = Vouch::buddies_in_set(addr);
+      let (is_found, _idx) = Vector::index_of(&buddies, &voucher);
+      assert(is_found, 100103);
+
+      // check the node has been mining before unjailing.
+      assert(TowerState::node_above_thresh(addr), 100104);
```

Comment:
why is the mining check there? It will be done at epoch boundary right? One voucher might want to unjail while the validator has not submitted 8 proofs yet.

Reply:
Yeah I suppose that's a reasonable point. But we don't have this check actually at the epoch boundary, so we would need it here.